commit 1580457f32d1bf0b51ab58a94575e1b0742f470e
Author: David James <davidjames@chromium.org>
Date:   Mon May 9 13:00:11 2011 -0700

    Update Portage to support fixing *.la and *.pc files.
    
    *.la and *.pc files need to be fixed up in order to ensure packages link
    against the right library when we try to build / install to different ROOTs.
    
    BUG=chromium-os:11316
    TEST=Compile with build_packages --nousepkg
    
    Review URL: http://codereview.chromium.org/6370011
    
    Change-Id: I99db793c3c0dc3fab7646a7fbb258fdc40b086a9
    (cherry picked from commit 0448eb511f63b2a1a99e724fe72764473b4958a4)
    
    Conflicts:
    
    	pym/portage/const.py
    	pym/portage/dbapi/vartree.py

diff --git a/pym/portage/const.py b/pym/portage/const.py
index ed8a460..9cd0daf 100644
--- a/pym/portage/const.py
+++ b/pym/portage/const.py
@@ -93,8 +93,8 @@ SUPPORTED_FEATURES       = frozenset([
                            "installsources", "keeptemp", "keepwork", "fixlafiles", "lmirror",
                            "metadata-transfer", "mirror", "multilib-strict", "news",
                            "noauto", "noclean", "nodoc", "noinfo", "noman",
-                           "nostrip", "notitles", "parallel-fetch", "parallel-install",
-                           "parse-eapi-ebuild-head",
+                           "nostrip", "notitles", "no-env-update", "parallel-fetch",
+                           "parallel-install", "parse-eapi-ebuild-head",
                            "prelink-checksums", "preserve-libs",
                            "protect-owned", "python-trace", "sandbox",
                            "selinux", "sesandbox", "severe", "sfperms",
diff --git a/pym/portage/dbapi/vartree.py b/pym/portage/dbapi/vartree.py
index 5b1bb51..ac2a0ce 100644
--- a/pym/portage/dbapi/vartree.py
+++ b/pym/portage/dbapi/vartree.py
@@ -60,6 +60,7 @@ from _emerge.PollScheduler import PollScheduler
 from _emerge.MiscFunctionsProcess import MiscFunctionsProcess
 
 import codecs
+import fileinput
 import gc
 import re, shutil, stat, errno, subprocess
 import logging
@@ -1751,17 +1752,18 @@ class dblink(object):
 		else:
 			self.settings.pop("PORTAGE_LOG_FILE", None)
 
-		# Lock the config memory file to prevent symlink creation
-		# in merge_contents from overlapping with env-update.
-		conf_mem_file = os.path.join(self._eroot, CONFIG_MEMORY_FILE)
-		conf_mem_lock = lockfile(conf_mem_file)
-		try:
-			env_update(target_root=self.settings['ROOT'],
-				prev_mtimes=ldpath_mtimes,
-				contents=contents, env=self.settings.environ(),
-				writemsg_level=self._display_merge)
-		finally:
-			unlockfile(conf_mem_lock)
+		if 'no-env-update' not in self.settings.features:
+			# Lock the config memory file to prevent symlink creation
+			# in merge_contents from overlapping with env-update.
+			conf_mem_file = os.path.join(self._eroot, CONFIG_MEMORY_FILE)
+			conf_mem_lock = lockfile(conf_mem_file)
+			try:
+				env_update(target_root=self.settings['ROOT'],
+					prev_mtimes=ldpath_mtimes,
+					contents=contents, env=self.settings.environ(),
+					writemsg_level=self._display_merge)
+			finally:
+				unlockfile(conf_mem_lock)
 
 		return os.EX_OK
 
@@ -3465,23 +3467,37 @@ class dblink(object):
 			showMessage(_("!!! FAILED postinst: ")+str(a)+"\n",
 				level=logging.ERROR, noiselevel=-1)
 
-		downgrade = False
-		for v in otherversions:
-			if pkgcmp(catpkgsplit(self.pkg)[1:], catpkgsplit(v)[1:]) < 0:
-				downgrade = True
+		if 'no-env-update' not in self.settings.features:
+			downgrade = False
+			for v in otherversions:
+				if pkgcmp(catpkgsplit(self.pkg)[1:], catpkgsplit(v)[1:]) < 0:
+					downgrade = True
 
-		# Lock the config memory file to prevent symlink creation
-		# in merge_contents from overlapping with env-update.
-		conf_mem_file = os.path.join(self._eroot, CONFIG_MEMORY_FILE)
-		conf_mem_lock = lockfile(conf_mem_file)
-		try:
-			#update environment settings, library paths. DO NOT change symlinks.
-			env_update(makelinks=(not downgrade),
-				target_root=self.settings['ROOT'], prev_mtimes=prev_mtimes,
-				contents=contents, env=self.settings.environ(),
-				writemsg_level=self._display_merge)
-		finally:
-			unlockfile(conf_mem_lock)
+			# Lock the config memory file to prevent symlink creation
+			# in merge_contents from overlapping with env-update.
+			conf_mem_file = os.path.join(self._eroot, CONFIG_MEMORY_FILE)
+			conf_mem_lock = lockfile(conf_mem_file)
+			try:
+				#update environment settings, library paths. DO NOT change symlinks.
+				env_update(makelinks=(not downgrade),
+					target_root=self.settings['ROOT'], prev_mtimes=prev_mtimes,
+					contents=contents, env=self.settings.environ(),
+					writemsg_level=self._display_merge)
+			finally:
+				unlockfile(conf_mem_lock)
+
+		# Fix *.{la,pc} files to point to libs in target_root, if they
+		# don't do so already.
+		re_root = self.settings["ROOT"].strip("/")
+		if re_root:
+			fix_files = []
+			for path in contents:
+				if path.endswith(".la") or path.endswith(".pc"):
+					if os.path.exists(path): fix_files.append(path)
+			if fix_files:
+				pat = re.compile(r"([' =](?:-[IL])?/)(usr|lib|opt)")
+				for line in fileinput.input(fix_files, inplace=1):
+					sys.stdout.write(pat.sub(r"\1%s/\2" % re_root, line))
 
 		# For gcc upgrades, preserved libs have to be removed after the
 		# the library path has been updated.
diff --git a/pym/portage/package/ebuild/doebuild.py b/pym/portage/package/ebuild/doebuild.py
index f83a1e3..0b5bc5f 100644
--- a/pym/portage/package/ebuild/doebuild.py
+++ b/pym/portage/package/ebuild/doebuild.py
@@ -4,6 +4,7 @@
 __all__ = ['doebuild', 'doebuild_environment', 'spawn', 'spawnebuild']
 
 import codecs
+import fileinput
 import gzip
 from itertools import chain
 import logging
@@ -1438,6 +1439,7 @@ def _post_src_install_uid_fix(mysettings, out):
 
 	destdir = mysettings["D"]
 	unicode_errors = []
+	fix_files = []
 
 	while True:
 
@@ -1526,10 +1528,12 @@ def _post_src_install_uid_fix(mysettings, out):
 							new_contents, mode='wb')
 
 				mystat = os.lstat(fpath)
-				if stat.S_ISREG(mystat.st_mode) and \
-					mystat.st_ino not in counted_inodes:
-					counted_inodes.add(mystat.st_ino)
-					size += mystat.st_size
+				if stat.S_ISREG(mystat.st_mode):
+					if fname.endswith(".pc") or fname.endswith(".la"):
+						fix_files.append(fpath)
+					if mystat.st_ino not in counted_inodes:
+						counted_inodes.add(mystat.st_ino)
+						size += mystat.st_size
 				if mystat.st_uid != portage_uid and \
 					mystat.st_gid != portage_gid:
 					continue
@@ -1597,6 +1601,14 @@ def _post_src_install_uid_fix(mysettings, out):
 			mode='w', encoding=_encodings['repo.content'],
 			errors='strict').write(v + '\n')
 
+	re_root = mysettings["ROOT"].strip("/")
+	if fix_files and re_root:
+		# Replace references to our sysroot with references to "/" in binpkg.
+		# Sysroot will be re-appended when the package is installed.
+		pat = re.compile(r"([' =](-[IL])?/)%s/" % re.escape(re_root))
+		for line in fileinput.input(fix_files, inplace=1):
+			sys.stdout.write(pat.sub(r"\1", line))
+
 	if bsd_chflags:
 		# Restore all of the flags saved above.
 		os.system("mtree -e -p %s -U -k flags < %s > /dev/null" % \
